onnxruntime/core/framework/session_state_utils.cc

```diff
@@ -31,58 +31,54 @@ namespace onnxruntime {
 namespace session_state_utils {
 
 static common::Status DeserializeTensorProto(const Env& env, const std::basic_string<PATH_CHAR_TYPE>& proto_path,
-                                             const ONNX_NAMESPACE::TensorProto& tensor_proto, const MemBuffer& m,
-                                             const OrtMemoryInfo& default_cpu_memory_info, OrtValue& ort_value,
-                                             OrtCallback& deleter,
-                                             const DataTransferManager& data_transfer_mgr) {
-  const OrtMemoryInfo& alloc_info = m.GetAllocInfo();
-  if (strcmp(alloc_info.name, CPU) == 0 || alloc_info.mem_type == OrtMemTypeCPUOutput) {
-    // deserialize directly to CPU tensor
-    return utils::TensorProtoToMLValue(env, proto_path.c_str(), tensor_proto, m, ort_value, deleter);
+                                             const ONNX_NAMESPACE::TensorProto& tensor_proto, const MemBuffer* m,
+                                             const AllocatorPtr& alloc, const AllocatorPtr& default_cpu_alloc,
+                                             OrtValue& ort_value, const DataTransferManager& data_transfer_mgr) {
+  if (bool(alloc) == (m != nullptr)) {
+    return Status(common::ONNXRUNTIME, common::INVALID_ARGUMENT, 
+        "DeserializeTensorProto() takes either pre-allocated buffer or an allocator!");
   }
 
-  // deserialize and copy. In the copy stage, it won't check if the buffer has enough room.
-  // The result tensor won't need a deleter because:
-  // 1. It mustn't be a string tensor
-  // 2. The memory is not memory-mapped.
-  deleter.f = nullptr;
-  deleter.param = nullptr;
-  if (tensor_proto.data_type() == ONNX_NAMESPACE::TensorProto_DataType_STRING) {
-    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "string tensor is not supported for copying between allocators");
+  // Get shape and type of the tensor, and allocate the empty tensor
+  TensorShape tensor_shape{utils::GetTensorShapeFromTensorProto(tensor_proto)};
+  const DataTypeImpl* const type = DataTypeImpl::TensorTypeFromONNXEnum(tensor_proto.data_type())->GetElementType();
+  std::unique_ptr<Tensor> p_tensor;
+  if (m != nullptr) {
+    p_tensor = onnxruntime::make_unique<Tensor>(type, tensor_shape, m->GetBuffer(), m->GetAllocInfo());
+    if (m->GetLen() < p_tensor->SizeInBytes()) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Internal error. The preallocated buffer is too small. Requires ",
+                             p_tensor->SizeInBytes(), ", Got ", m->GetLen());
+    }
+  } else {
+    // tensor constructor should give us enough buffer size based on type and shape
+    p_tensor = onnxruntime::make_unique<Tensor>(type, tensor_shape, alloc);
   }
```

Comment:
Nice!

Reply:
But indeed this is the function that we want to avoid using.  I put a comment there saying it is deprecated.  Because if you manually manage the buffers, even though the code will be more complicated and hard to read but it will be also more efficient.  If you let the Tensor class creates the buffers, then you will have double copies of the buffer at some time which increases the peak memory usage.